core/valor_core/classification.py

```diff
@@ -0,0 +1,1291 @@
+                    chain_df["n_true_positives"].isnull(),
+                    ((chain_df["n"] - chain_df["n_true_positives"]) == 0),
+                ],
+                [1, 1],
+                default=chain_df["trap_area"],
+            )
+        )
+    )
+
+    # take the average across grouper keys
+    average_across_label_keys = summed_trap_areas_per_label_value.groupby(
+        "label_key", as_index=False
+    )["trap_area"].mean()
+
+    return [
+        metrics.ROCAUCMetric(
+            label_key=values["label_key"], value=values["trap_area"]
+        )
+        for _, values in average_across_label_keys.iterrows()
+    ]
+
+
+def _add_samples_to_dataframe(
+    pr_curve_counts_df: pd.DataFrame,
+    pr_calc_df: pd.DataFrame,
+    max_examples: int,
+    flag_column: str,
+) -> pd.DataFrame:
+    """Efficienctly gather samples for a given flag."""
+
+    sample_df = pd.concat(
+        [
+            pr_calc_df[pr_calc_df[flag_column]]
+            .groupby(
+                [
+                    "label_key",
+                    "label_value_gt",
+                    "confidence_threshold",
+                ],
+                as_index=False,
+            )[["datum_uid"]]
+            .agg(lambda x: tuple(x.head(max_examples)))
+            .rename(columns={"label_value_gt": "label_value"}),
+            pr_calc_df[pr_calc_df[flag_column]]
+            .groupby(
+                [
+                    "label_key",
+                    "label_value_pd",
+                    "confidence_threshold",
+                ],
+                as_index=False,
+            )[["datum_uid"]]
+            .agg(lambda x: tuple(x.head(max_examples)))
+            .rename(columns={"label_value_pd": "label_value"}),
+        ],
+        axis=0,
+    ).drop_duplicates()
+
+    if not sample_df.empty:
+        sample_df[f"{flag_column}_samples"] = sample_df.apply(
+            lambda row: set(zip(*row[["datum_uid"]])),  # type: ignore - pandas typing error
+            axis=1,
+        )
+
+        pr_curve_counts_df = pr_curve_counts_df.merge(
+            sample_df[
+                [
+                    "label_key",
+                    "label_value",
+                    "confidence_threshold",
+                    f"{flag_column}_samples",
+                ]
+            ],
+            on=["label_key", "label_value", "confidence_threshold"],
+            how="outer",
+        )
+        pr_curve_counts_df[f"{flag_column}_samples"] = pr_curve_counts_df[
+            f"{flag_column}_samples"
+        ].apply(lambda x: list(x) if isinstance(x, set) else list())
+
+    else:
+        pr_curve_counts_df[f"{flag_column}_samples"] = [
+            list() for _ in range(len(pr_curve_counts_df))
+        ]
+
+    return pr_curve_counts_df
+
+
+def _calculate_pr_curves(
+    prediction_df: pd.DataFrame,
+    groundtruth_df: pd.DataFrame,
+    metrics_to_return: list,
+    pr_curve_max_examples: int,
+) -> list[metrics.PrecisionRecallCurve]:
```

Comment:
Might make sense in this function to use some kind of "label ID" instead of a string `label_value`.  The strings are redundant and can use lots of memory if they're big.  Map to an integer and then back again at the very end?

Reply:
label tuples are significantly easier to debug than label IDs. I don't think the trade-off is worth it here as these strings are usually very small

Reply:
@rsbowman-striveworks I agree with your suggestion but this is something we would likely implement in a follow-up PR. Would you be able to add this to your memory ticket?